libraries/textcomposer/impl/src/main/kotlin/io/element/android/libraries/textcomposer/TextComposerState.kt

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright (c) 2023 New Vector Ltd
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.element.android.libraries.textcomposer
+
+import androidx.compose.runtime.Composable
+import io.element.android.wysiwyg.compose.RichTextEditorState
+import io.element.android.wysiwyg.compose.rememberRichTextEditorState
+
+data class TextComposerState constructor(
+    internal val inner: InnerTextComposerState
+) : InnerTextComposerState {
+    override val messageHtml: String
+        get() = inner.messageHtml
+
+    override val messageMarkdown: String
+        get() = inner.messageMarkdown
+
+    override val hasFocus: Boolean
+        get() = inner.hasFocus
+
+    override val lineCount: Int
+        get() = inner.lineCount
+
+    override val canSendMessage: Boolean
+        get() = messageHtml.isNotEmpty()
+
+    override fun setHtml(value: String) {
+        inner.setHtml(value)
+    }
+
+    override fun requestFocus(): Boolean =
+        inner.requestFocus()
+}
+
+interface InnerTextComposerState {
+    val messageHtml: String
+    val messageMarkdown: String
+    val hasFocus: Boolean
+    val lineCount: Int
+    val canSendMessage: Boolean
+
+    fun setHtml(value: String)
+    fun requestFocus(): Boolean
+}
+
+data class RealInnerTextComposerState internal constructor(
+    internal val richTextEditorState: RichTextEditorState
+) : InnerTextComposerState {
+    override val messageHtml: String
+        get() = richTextEditorState.messageHtml
+
+    override val messageMarkdown: String
+        get() = richTextEditorState.messageMarkdown
+
+    override val hasFocus: Boolean
+        get() = richTextEditorState.hasFocus
+
+    override val lineCount: Int
+        get() = richTextEditorState.lineCount
+
+    override val canSendMessage: Boolean
+        get() = messageHtml.isNotEmpty()
+
+    override fun setHtml(value: String) {
+        richTextEditorState.setHtml(value)
+    }
+
+    override fun requestFocus(): Boolean =
+        richTextEditorState.requestFocus()
+}
+
+@Composable
+fun rememberTextComposerState(): TextComposerState {
+    val richTextEditorState = rememberRichTextEditorState()
+
+    return TextComposerState(RealInnerTextComposerState(richTextEditorState))
+}
+
+fun previewTextComposerState(text: String): TextComposerState {
+    val richTextEditorState = RichTextEditorState.createForLocalInspectionMode(text)
+
+    return TextComposerState(RealInnerTextComposerState(richTextEditorState))
+}
```

Comment:
This whole file looks a bit overcomplicated. I would just remove it. Can't we directly use `rememberRichTextEditorState("")` when we need a `RichTextEditorState` (since its constructor is internal :/).

`RichTextEditorState` constructor should be public for the testing reason IMO.

Reply:
Agreed, the original abstraction aimed to limit exposure of the rich text editor beyond this module and facilitate testing, (particularly `setHtml` using a [fake implementation](https://github.com/vector-im/element-x-android/pull/1172/files#diff-d8e85d6e3a283eb3dfb95bfc13f42f48b7d6144d8b3de9ab5874e8e2896ad9d7R40)). But it has definitely led to too much complexity. I will get rid of it and work directly with the RTE state.

I've opened a PR to address this issue in the library: https://github.com/matrix-org/matrix-rich-text-editor/pull/796

Reply:
Thanks!